src/transformers/models/llama/modeling_llama.py

```diff
@@ -89,11 +89,11 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
     bsz, src_len = mask.size()
     tgt_len = tgt_len if tgt_len is not None else src_len
 
-    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
+    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).transpose(-2,-1).to(dtype)
```

Comment:
This is copied from bart, changing this would need some justifications. You are transposing target ans source length, so the output mask is no longer `[bsz, 1, tgt_seq_len, src_seq_len]`  but `[bsz, 1, src_seq_len, tgt_seq_len]`

Reply:
As I understand, `tgt_seq_len != src_seq_len` happens when it incurs cross attention.
Suppose the source sequence length is `s_1` and target sequence length is `s_2`, then we have Q, K, V shape as:
- Q: s_1 * h
- K: s_2 * h
- V: s_2 * h

Then according to formula: `(Q \dot K^T + attention_mask) \dot V`, we know that the shape of attention mask should be `s_1 * s_2`. That is, with batch size included the shape should be `(bsz, 1, src_seq_len, tgt_seq_len)`.

It would be great if we can include some unit tests to verify the change here. But I checked the unit test [1] and did not find how to verify this.

Or any suggestions on this?

[1] https://github.com/huggingface/transformers/blob/897a826d830e8b1e03eb482b165b5d88a7a08d5f/tests/models/llama/test_modeling_llama.py

Reply:
Hi, I'm not sure about this change. LLaMA does not support cross-attention - `tgt_seq_len != src_seq_len` occurs when `past_key_values` is present. If the original code was wrong, then we would get shape errors whenever `past_key_values` is passed to `LLaMA`. Since we don't see those errors, I have to assume that the original code is correct and this transpose is wrong!

Reply:
Hi  @Rocketknight1 , thanks for the review. You are right and `tgt_seq_len != src_seq_len` happens when `past_key_values` is present.

The reason that I created this PR is that I found the output of using flash attention or not leads to different attention outputs [1][2] when padding mask is not null. Unfortunately the solution before is incorrect when `past_key_values` is present.

I have refined the solution in this PR and now the attention outputs of using flash attention or not leads to the same output [1][2]. Please take a look.

BTW, you can use [3] to verify the bug I found.

[1] https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L398
[2] https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L498
[3] 
```
import random
import torch

torch_device = "cuda" if torch.cuda.is_available() else "cpu"

from transformers import LlamaConfig, is_torch_available, set_seed
from transformers import LlamaForCausalLM, LlamaForSequenceClassification, LlamaModel, LlamaTokenizer

class LlamaTest:
    def __init__(
        self,
        is_training=True,
        use_input_mask=True,
        use_token_type_ids=False,
        use_labels=True,
        vocab_size=99,
        hidden_size=32,
        num_hidden_layers=1,
        num_attention_heads=4,
        intermediate_size=37,
        hidden_act="gelu",
        hidden_dropout_prob=0.1,
        attention_probs_dropout_prob=0.1,
        max_position_embeddings=512,
        type_vocab_size=16,
        type_sequence_label_size=2,
        initializer_range=0.02,
        num_labels=3,
        num_choices=4,
        pad_token_id=0,
        scope=None,
    ):
        self.is_training = is_training
        self.use_input_mask = use_input_mask
        self.use_token_type_ids = use_token_type_ids
        self.use_labels = use_labels
        self.vocab_size = vocab_size
        self.hidden_size = hidden_size
        self.num_hidden_layers = num_hidden_layers
        self.num_attention_heads = num_attention_heads
        self.intermediate_size = intermediate_size
        self.hidden_act = hidden_act
        self.hidden_dropout_prob = hidden_dropout_prob
        self.attention_probs_dropout_prob = attention_probs_dropout_prob
        self.max_position_embeddings = max_position_embeddings
        self.type_vocab_size = type_vocab_size
        self.type_sequence_label_size = type_sequence_label_size
        self.initializer_range = initializer_range
        self.num_labels = num_labels
        self.num_choices = num_choices
        self.pad_token_id = pad_token_id
        self.scope = scope

    def get_config(self, use_fa=False):
        return LlamaConfig(
            vocab_size=self.vocab_size,
            hidden_size=self.hidden_size,
            num_hidden_layers=self.num_hidden_layers,
            num_attention_heads=self.num_attention_heads,
            intermediate_size=self.intermediate_size,
            hidden_act=self.hidden_act,
            hidden_dropout_prob=self.hidden_dropout_prob,
            attention_probs_dropout_prob=self.attention_probs_dropout_prob,
            max_position_embeddings=self.max_position_embeddings,
            type_vocab_size=self.type_vocab_size,
            is_decoder=False,
            initializer_range=self.initializer_range,
            pad_token_id=self.pad_token_id,
            _flash_attn_2_enabled=use_fa
        )

def convert_to_bf16(model):
    for param in model.parameters():
        param.data = param.data.bfloat16()
    model = model.bfloat16()
    return model


if __name__ == "__main__":
    batch_size = 1
    seq_length=8
    llama_test = LlamaTest()
    llama_config = llama_test.get_config(False)
    model = LlamaModel(llama_config)
    model = convert_to_bf16(model)
    model.to(torch_device)
    #torch.save(model.state_dict(), "llama_model.pth")
    model.load_state_dict(torch.load("llama_model.pth"))
    model.eval()
    input_mask = torch.tensor(data=[1,1,1,1,1,0,0,0], dtype=torch.long, device=torch_device).view([batch_size, seq_length]).contiguous()
    input_ids = torch.tensor(data=[1,2,3,4,5,6,7,8], dtype=torch.long, device=torch_device).view([batch_size, seq_length]).contiguous()
    print(input_ids)
    print(input_mask)
    model(input_ids, attention_mask=input_mask)
```

Reply:
Hi @zhipeng93 - there's actually been a change here that I wasn't expecting! Because of the issues we had with attention masks, a new class `AttentionMaskConverter` has just been added - you can see it [here](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_attn_mask_utils.py).

If you rebase this PR, it should be possible to use that class to handle the attention mask computations. I believe this is now the recommended way to handle attention masks, and should fix a lot of these problems.

Can you try it and let me know if it works, and if you have any difficulties, let me know?

Reply:
Hi @Rocketknight1 , I have rebased the main branch and the above script still produces the same result --- the attention output of transformers is inconsistent with flash-attention2.

Please take a look at the updated PR.